pandas/core/dtypes/concat.py

```diff
@@ -498,9 +498,16 @@ def _concat_index_asobject(to_concat, name=None):
     attribs = self._get_attributes_dict()
     attribs['name'] = name
 
-    to_concat = [x._values if isinstance(x, Index) else x
-                 for x in to_concat]
-    return self._shallow_copy_with_infer(np.concatenate(to_concat), **attribs)
+    arrays = []
+    for x in to_concat:
+        if is_categorical_dtype(x):
```

Comment:
why is this a special case?

Reply:
@TomAugspurger don't we want ``np.asarray(x, dtype=object)`` for *all* extension dtypes in this case? 

So for Period and Interval `_values`  (used some lines below for Index types) currently still returns object array, but I think the purpose is that those will also return their own Array class, so using ``np.asarray(x, dtype=object)`` will do the same now and is more future proof?